x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongScriptDocValues.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.unsignedlong;
+
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.util.ArrayUtil;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.search.DocValueFormat;
+
+import java.io.IOException;
+
+public class UnsignedLongScriptDocValues extends ScriptDocValues<Number> {
+    private final SortedNumericDocValues in;
+    private Number[] values = new Number[0];
+    private int count;
+
+    /**
+     * Standard constructor.
+     */
+    public UnsignedLongScriptDocValues(SortedNumericDocValues in) {
+        this.in = in;
+    }
+
+    @Override
+    public void setNextDocId(int docId) throws IOException {
+        if (in.advanceExact(docId)) {
+            resize(in.docValueCount());
+            for (int i = 0; i < count; i++) {
+                values[i] = (Number) DocValueFormat.UNSIGNED_LONG_SHIFTED.format(in.nextValue());
```

Comment:
We'd need to provide the doc value format but that could be done in the `ScriptDocValues.Long` directly ?

Reply:
@jimczi  Thanks for the feedback. I am not very clear what you are suggesting here. Are you suggesting to use `ScriptDocValues.Longs` instead of creating a new class?   If yes, I have opted to create a new class, as `ScriptDocValues.Longs` hold values in a array  `long[] values`, while we need an array of `Number[] values` as sometimes produced values can be of type `long` and other times of type `BigInteger`.

Reply:
So today `ScriptDocValues.Long` returns a `Long` but it could return a `Number` instead and rely on a provided doc values format to do the conversion to a number. In the normal case we would conver to a `Long` but if the provided doc values format is the `unsigned_long` then it can also return a `BigInteger` in some cases. 
I am not sure it's a good idea so feel free to keep the code separated if you prefer. In any case I don't think we need to rely on a `Number[]` array. We can store the unsigned long in a plain `long[]` array and do the conversion to `Number` lazily when `getValue` or `get(int)` are called.

Reply:
@jimczi  Thanks Jim, this is a good idea. I explored using `Longs` but it turned out it involves more modifications as `Longs::getValue` expects `long`.  Instead, I followed your 2nd idea to used `long[]` for values in `UnsignedLongScriptDocValues.java `  and this is addressed in 2b567c96.

Reply:
++, thanks